source/module_hsolver/diago_cusolver.cpp

```diff
@@ -135,45 +135,59 @@ namespace hsolver
         ModuleBase::timer::tick("DiagoCusolver", "cusolver");
 
 #ifdef __MPI
-        // gather matrices from processes to root process
-        gatherMatrix(myid, root_proc, h_mat, h_mat_g);
-        gatherMatrix(myid, root_proc, s_mat, s_mat_g);
+        if(num_procs > 1)
+        {
+            // gather matrices from processes to root process
+            gatherMatrix(myid, root_proc, h_mat, h_mat_g);
+            gatherMatrix(myid, root_proc, s_mat, s_mat_g);
+        }
 #endif
 
         // Call the dense diagonalization routine
 #ifdef __MPI
-        MPI_Barrier(MPI_COMM_WORLD);
-        if (myid == root_proc)
+        if(num_procs > 1)
         {
-            psi_g = new T[h_mat_g.row * h_mat_g.col];
-            this->dc.Dngvd(h_mat_g.col, h_mat_g.row, h_mat_g.p, s_mat_g.p, eigen.data(), psi_g);
+            MPI_Barrier(MPI_COMM_WORLD);
+            if (myid == root_proc)
+            {
+                psi_g = new T[h_mat_g.row * h_mat_g.col];
```

Comment:
is it possible to use STL vector instead of new?

Reply:
`psi_g` is used in Dngvd. It would require modifying the function interface to change the variable type, so altering the variable type seems impractical.
![image](https://github.com/deepmodeling/abacus-develop/assets/153698752/4f2ea4d3-18e0-4e68-8c9f-dd1167a7461b)

Reply:
try harder to resolve the issue